sklearn/decomposition/tests/test_pca.py

```diff
@@ -260,7 +260,46 @@ def test_infer_dim_3():
     pca = PCA(n_components=p)
     pca.fit(X)
     spect = pca.explained_variance_
-    assert_greater(_infer_dimension_(spect, n, p), 2)
+    assert_greater(_infer_dimension(spect, n, p), 2)
+
+
+def test_infer_dim_bad_spec():
+    """
+    Test that we deal with a spectrum that drops to near zero
+    see issue https://github.com/scikit-learn/scikit-learn/issues/4441
+    """
+    spectrum = np.array([1, 1e-30, 1e-30, 1e-30])
+    n_samples = 10
+    n_features = 5
+    ret = _infer_dimension(spectrum, n_samples, n_features)
+    assert_equal(ret, 0)
+
+
+def test_assess_dimension_tiny_eigenvals():
+    """
+    Test that we deal with tiny eigenvalues appropriately when
+    `mle` inferring `n_components`
+    see issue https://github.com/scikit-learn/scikit-learn/issues/4441
+    """
+    spectrum = np.array([1, 1e-30, 1e-30, 1e-30])
+    n_samples = 10
+    n_features = 5
+    rank = 4
+    ret = _assess_dimension(spectrum, rank, n_samples, n_features)
+    assert_equal(ret, -np.inf)
+
+
+def test_infer_dim_mle():
+    """
+    Test that we deal with tiny eigenvalues appropriately when
+    `mle` inferring `n_components` with a pathalogical `X` dastaset
+    see issue https://github.com/scikit-learn/scikit-learn/issues/4441
+    """
+    X, _ = datasets.make_classification(n_informative=1, n_repeated=18,
+                                        n_redundant=1, n_clusters_per_class=1,
+                                        random_state=20150609)
+    pca = PCA(n_components='mle').fit(X)
+    assert_equal(pca.n_components_, 0)
```

Comment:
Is this related to the off-by-one error that @alexis-mignon talked about in the original comment on 24 Mar?